core/src/generic.rs

```diff
@@ -67,47 +77,71 @@
 /// # }
 /// ```
 pub trait Generic {
-    /// The generic representation type
+    /// The generic representation type.
     type Repr;
 
-    /// Go from something to Repr
+    /// Convert an object to its representation type `Repr`.
     fn into(self) -> Self::Repr;
 
-    /// Go from Repr to something
-    fn from(r: Self::Repr) -> Self;
+    /// Convert an object's representation type `Repr` to the object's type.
+    fn from(repr: Self::Repr) -> Self;
 
-    /// From one type to another using a type with a compatible generic representation
-    fn convert_from<A>(a: A) -> Self
+    /// Convert an object to a another type provided that they have
+    /// a compatible representation type.
+    fn convert_from<Src>(src: Src) -> Self
     where
-        A: Generic<Repr = Self::Repr>,
+        Src: Generic<Repr = Self::Repr>,
         Self: Sized,
     {
-        let repr = <A as Generic>::into(a);
+        let repr = <Src as Generic>::into(src);
         <Self as Generic>::from(repr)
     }
+
+    /// Maps the given object of type `Self` by first transforming it to
+    /// the representation type, then applying a `mapper` function
+    /// on the representation type and finally transforming it back to
+    /// an object of type `Self`.
+    fn map_repr<Repr, Mapper>(self, mapper: Mapper) -> Self
+    where
+        Self: Generic<Repr = Repr> + Sized,
+        Mapper: FnOnce(Repr) -> Repr
+    {
+        Self::from(mapper(self.into()))
+    }
 }
 
-/// Given a generic Representation of an A, returns A
-pub fn from_generic<A, Repr>(gen: Repr) -> A
+/// Given a generic representation `Repr` of a `Dst`, returns `Dst`.
+pub fn from_generic<Dst, Repr>(repr: Repr) -> Dst
 where
-    A: Generic<Repr = Repr>,
+    Dst: Generic<Repr = Repr>,
 {
-    <A as Generic>::from(gen)
+    <Dst as Generic>::from(repr)
 }
 
-/// Given an A, returns its generic Representation
-pub fn into_generic<A, Repr>(a: A) -> Repr
+/// Given a object of type `Src`, returns its generic representation `Repr`.
+pub fn into_generic<Src, Repr>(src: Src) -> Repr
```

Comment:
I'm just noticing this now, but maybe Repr shouldn't be in the type parameter list here (since it's not necessary, and the type parameter list is public API).